Mathlib/Data/List/Shortlex.lean

```diff
@@ -0,0 +1,285 @@
+/-
+Copyright (c) 2024 Hannah Fechtner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hannah Fechtner
+-/
+
+import Mathlib.Data.List.Lex
+import Mathlib.Tactic.Linarith
+
+/-!
+# Shortlex ordering of lists.
+
+Given a relation `r` on `α`, the shortlex order on `List α` is defined by `L < M` iff
+* `L.length < M.length`
+* `L.length = M.length` and `L < M` under the lexicographic ordering over `r` on lists
+
+## Main results
+
+We show that if `r` is well-founded, so too is the shortlex order over `r`
+
+## See also
+
+Related files are:
+* `Mathlib.Data.List.Lex`: Lexicographic order on `List α`.
+* `Mathlib.Data.DFinsupp.WellFounded`: Well-foundedness of lexicographic orders on `DFinsupp` and
+  `Pi`.
+-/
+
+/-! ### shortlex ordering -/
+
+namespace List
+
+/-- Given a relation `r` on `α`, the shortlex order on `List α`, for which
+`[a0, ..., an] < [b0, ..., b_k]` if `n < k` or `n = k` and `[a0, ..., an] < [b0, ..., bk]`
+under the lexicographic order induced by `r`. -/
+def Shortlex {α : Type*} (r : α → α → Prop) : List α → List α → Prop :=
+  fun a b => Prod.Lex (fun n1 n2 => n1 < n2) (fun a b => List.Lex r a b) (a.length, a) (b.length, b)
+
+namespace Shortlex
+
+variable {α : Type*} {r : α → α → Prop}
+
+/-- If a list `s` is shorter than a list `t`, then `s` is smaller than `t` under any shortlex
+order -/
+theorem of_length_lt {s t : List α} (h : s.length < t.length) : Shortlex r s t :=
+  Prod.Lex.left _ _ h
+
+/-- If two lists `s` and `t` have the same length, `s` is smaller than `t` under the shortlex order
+over a relation `r`  when `s` is smaller than `t` under the lexicographic order over `r`-/
+theorem of_lex {s t : List α} (h : s.length = t.length) (h2 : List.Lex r s t) :
+    Shortlex r s t := by
+  unfold Shortlex
+  rw [h]
+  exact Prod.Lex.right _ h2
+
+/-- If two lists `s` and `t` have the same length, `s` is smaller than `t` under the shortlex order
+over a relation `r` exactly when `s` is smaller than `t` under the lexicographic order over `r`-/
+theorem of_lex_iff {s t : List α} (h : s.length = t.length) : Shortlex r s t ↔
+    List.Lex r s t := by
```

Comment:
```suggestion
theorem _root_.List.shortlex_iff_lex {s t : List α} (h : s.length = t.length) :
    Shortlex r s t ↔ List.Lex r s t := by
```
The "of" naming is about implication only, not `iff`. You could add `of_length_eq` on the end here if you wanted, but I think it's fine to keep it short as there's no ambiguity.